### DISCUSSION TITLE
problem with DCP rules

### DISCUSSION THREAD

**Discussion by Pacasas** [cvxpy/cvxpy#2066] [Q&A] [answered]
Hi, i am fairly new to convex optimization, however i am trying to create a type of LR using cvxpy that penalizes some very specific type of fairness.  however i get that the it does not follow DCP rules and I have checked the functions and i cant seem to find the source of that error, (X01 and X11 are basically X where some of the rows have been turnt into 0s ) and etar 0 and 1 are scalars

```py
h = X @ beta + b
h0=X_01 @ beta + b
h1=X_11 @ beta + b

log_likelihood0 = cp.sum(
    cp.multiply(y, h) - cp.logistic(h) +
    cp.multiply(etar0, cp.logistic(h0)) -
    cp.multiply(etar1, cp.logistic(h1))
)
log_likelihood1 = cp.sum(
    cp.multiply(y, h) - cp.logistic(h) +
    cp.multiply(etar1, cp.logistic(h1)) -
    cp.multiply(etar0, cp.logistic(h0))
)

cons = [log_likelihood1 - t <= 0, log_likelihood0 - t <= 0]
cp.Problem(cp.Minimize(t), cons)
```

**Comment by SteveDiamond** [comment, marked as answer]
``logistic`` of a linear function is convex. You have ``cp.multiply(etar0, cp.logistic(h0)) - cp.multiply(etar1, cp.logistic(h1))``. What are the signs of etar0 and etar1? Unless etar0 is all positive and etar1 is all negative (or vice versa) this is a DCP violation (convex - concave).

**Reply by Pacasas** [reply]
okay that makes sense, they are both positive (they are the penalization parameter times the inverse proportion of the class ) . I will try and work out how to solve the issue.

**Reply by Pacasas** [reply]
sorry, I did not realize, that can't be the only issue, as I have tried running it with etar0 and etar1 set to 0 and I get the same issue.

**Reply by SteveDiamond** [reply]
That's true, the constraint ``cp.sum(cp.multiply(y, h) - cp.logistic(h)) - t <= 0`` is concave <= 0, which is not DCP. It should be convex <= concave.

**Reply by Pacasas** [reply]
okay I think now i understand it, thank you very much, and i apologize if the question was too basic.

**Reply by SteveDiamond** [reply]
not at all!
